ledger/simulation/resources.go

```diff
@@ -0,0 +1,584 @@
+	}
+	return numBoxes
+}
+
+// GroupResourceAssignment calculates the additional resources that a transaction group could use,
+// and it tracks any referenced unnamed resources that fit within those limits.
+type GroupResourceAssignment struct {
+	// Resources specifies global resources for the entire group.
+	Resources ResourceAssignment
+
+	AssetHoldings             map[ledgercore.AccountAsset]struct{}
+	AppLocals                 map[ledgercore.AccountApp]struct{}
+	MaxCrossProductReferences int
+
+	// localTxnResources specifies local resources for each transaction in the group. This will only
+	// be populated if a top-level transaction executes AVM programs prior to v9 (when resource
+	// sharing was added).
+	localTxnResources []ResourceAssignment
+
+	startingBoxes int
+}
+
+func makeGroupResourceAssignment(txns []transactions.SignedTxnWithAD, proto *config.ConsensusParams) GroupResourceAssignment {
+	var startingBoxes int
+	var nonAppCalls int
+	localTxnResources := make([]ResourceAssignment, len(txns))
+	for i := range txns {
+		localTxnResources[i] = makeTxnResourceAssignment(&txns[i].Txn, proto)
+		startingBoxes += len(txns[i].Txn.Boxes)
+		if txns[i].Txn.Type != protocol.ApplicationCallTx {
+			nonAppCalls++
+		}
+	}
+	// Calculate the maximum number of cross-product resources that can be accessed by one app call
+	// under normal circumstances. This is calculated using the case of an app call with a full set
+	// of foreign apps. Including the app being called, there are (MaxAppTxnForeignApps + 1) apps,
+	// crossed with (MaxAppTxnForeignAssets + 2) accounts (the called app's account, the sender's
+	// account, and the foreign app accounts). We then subtract out the app local of sender's
+	// account and the called app, and each app local of an app and its own account, or
+	// (MaxAppTxnForeignApps + 2) references. So we end up with:
+	//
+	// (MaxAppTxnForeignApps + 1) * (MaxAppTxnForeignApps + 2) - (MaxAppTxnForeignApps + 2)
+	// <=> MaxAppTxnForeignApps^2 + 3*MaxAppTxnForeignApps + 2 - MaxAppTxnForeignApps - 2
+	// <=> MaxAppTxnForeignApps^2 + 2*MaxAppTxnForeignApps
+	// <=> MaxAppTxnForeignApps * (MaxAppTxnForeignApps + 2)
+	maxCrossProductsPerAppCall := proto.MaxAppTxnForeignApps * (proto.MaxAppTxnForeignApps + 2)
+	return GroupResourceAssignment{
+		Resources:                 makeGlobalResourceAssignment(localTxnResources, proto),
+		MaxCrossProductReferences: maxCrossProductsPerAppCall * (proto.MaxTxGroupSize - nonAppCalls),
+		localTxnResources:         localTxnResources,
+		startingBoxes:             startingBoxes,
+	}
+}
+
+func (a *GroupResourceAssignment) removePrivateFields() {
+	a.startingBoxes = 0
+	a.Resources.removePrivateFields()
+	for i := range a.localTxnResources {
+		a.localTxnResources[i].removePrivateFields()
+	}
+	a.localTxnResources = nil
+}
+
+func (a *GroupResourceAssignment) hasAccount(addr basics.Address, ep *logic.EvalParams, programVersion uint64, globalSharing bool, txnIndex int) bool {
```

Comment:
This is awfully nitpicky, and I'm totally ok if you you want your own convention in the `simulation` package, but I have been trying to always use `gi` for what you're calling `txnIndex` here.  It's obvious reasonable to think of the index as the index "of the transaction" _or_ as an index "into the group", but I've found it nice to settle on "group" and with consistency across the code base on that, I then feel ok with the very short name, just as we're ok with `ep` for EvalParams all over.

Reply:
Sure, I'll rename to `gi`